core/src/main/java/org/kiwix/kiwixmobile/core/page/bookmark/viewmodel/BookmarkViewModel.kt

```diff
@@ -1,110 +1,77 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
 package org.kiwix.kiwixmobile.core.page.bookmark.viewmodel
 
 import androidx.lifecycle.MutableLiveData
-import androidx.lifecycle.ViewModel
-import io.reactivex.disposables.CompositeDisposable
-import io.reactivex.processors.PublishProcessor
-import io.reactivex.schedulers.Schedulers
-import org.kiwix.kiwixmobile.core.base.SideEffect
 import org.kiwix.kiwixmobile.core.dao.NewBookmarksDao
 import org.kiwix.kiwixmobile.core.page.bookmark.adapter.BookmarkItem
-import org.kiwix.kiwixmobile.core.page.bookmark.viewmodel.effects.OpenBookmark
 import org.kiwix.kiwixmobile.core.page.bookmark.viewmodel.effects.ShowDeleteBookmarksDialog
 import org.kiwix.kiwixmobile.core.page.bookmark.viewmodel.effects.UpdateAllBookmarksPreference
 import org.kiwix.kiwixmobile.core.page.viewmodel.Action
-import org.kiwix.kiwixmobile.core.page.viewmodel.Action.Exit
-import org.kiwix.kiwixmobile.core.page.viewmodel.Action.ExitActionModeMenu
-import org.kiwix.kiwixmobile.core.page.viewmodel.Action.Filter
-import org.kiwix.kiwixmobile.core.page.viewmodel.Action.OnItemClick
-import org.kiwix.kiwixmobile.core.page.viewmodel.Action.OnItemLongClick
-import org.kiwix.kiwixmobile.core.page.viewmodel.Action.UpdatePages
-import org.kiwix.kiwixmobile.core.page.viewmodel.Action.UserClickedDeleteButton
-import org.kiwix.kiwixmobile.core.page.viewmodel.Action.UserClickedDeleteSelectedPages
-import org.kiwix.kiwixmobile.core.page.viewmodel.Action.UserClickedShowAllToggle
+import org.kiwix.kiwixmobile.core.page.viewmodel.PageViewModel
 import org.kiwix.kiwixmobile.core.reader.ZimReaderContainer
-import org.kiwix.kiwixmobile.core.search.viewmodel.effects.Finish
 import org.kiwix.kiwixmobile.core.utils.SharedPreferenceUtil
 import javax.inject.Inject
 
-const val TAG: String = "BookmarkViewModel"
-
 class BookmarkViewModel @Inject constructor(
-  private val bookmarksDao: NewBookmarksDao,
-  private val zimReaderContainer: ZimReaderContainer,
-  private val sharedPreferenceUtil: SharedPreferenceUtil
-) : ViewModel() {
-
-  val state = MutableLiveData<BookmarkState>().apply {
-    value =
-      BookmarkState(emptyList(), sharedPreferenceUtil.showBookmarksAllBooks, zimReaderContainer.id)
+  bookmarksDao: NewBookmarksDao,
+  override val zimReaderContainer: ZimReaderContainer,
+  override val sharedPreferenceUtil: SharedPreferenceUtil
+) : PageViewModel<BookmarkItem, BookmarkState>(bookmarksDao) {
+
+  override val state by lazy {
+    MutableLiveData<BookmarkState>().apply {
+      value =
+        BookmarkState(emptyList(), sharedPreferenceUtil.showHistoryAllBooks, zimReaderContainer.id)
+    }
   }
-  val effects = PublishProcessor.create<SideEffect<*>>()
-  val actions = PublishProcessor.create<Action>()
-  private val compositeDisposable = CompositeDisposable()
 
   init {
```

Comment:
this can be done in the parent class

Reply:
I suspect initalState is called before BookmarksViewModel can initalize the sharedpreferences object if init is moved to PageViewModel. This causes a null ptr exception when initialState() is called

```
  override fun initialState(): BookmarkState =
    BookmarkState(
      emptyList(),
      sharedPreferenceUtil.showBookmarksAllBooks, <------ null ptr exception here if init is moved to pageViewModel
      zimReaderContainer.id
    )
```

Reply:
Then move SharedPreferences to the constructor of PageViewModel

Reply:
Cool, I got it working. I tried doing this before and couldn't get it to work. I think what I missed is that this works:

```
  val actions = PublishProcessor.create<Action>()

  init {
    addDisposablesToCompositeDisposable()
  }

  private fun viewStateReducer(): Disposable =
    actions.map { reduce(it, state.value!!) }
      .subscribe(state::postValue, Throwable::printStackTrace)
```

but this does not:

```


  init {
    addDisposablesToCompositeDisposable()
  }

  private fun viewStateReducer(): Disposable =
    actions.map { reduce(it, state.value!!) }
      .subscribe(state::postValue, Throwable::printStackTrace)

  val actions = PublishProcessor.create<Action>()

```

Reply:
Yeah `init` blocks are called where they are declared at least relative to fields. You can in fact have multiple `init` blocks, see MainMenu for an example.

Reply:
Wow! Did not know that. It's pretty non-intuitive but works nicely in MainMenu.